contracts/libraries/ControllerV2.sol

```diff
@@ -0,0 +1,231 @@
+        if (leftAmount == 0) return;
+        ISubstitute(substitute).burn(leftAmount, to);
+    }
+
+    function _ensureBalance(address token, address user, uint256 amount) internal {
+        // TODO: consider to use SubstituteLibrary
+        address underlyingToken = ISubstitute(token).underlyingToken();
+        uint256 thisBalance = IERC20(token).balanceOf(address(this));
+        uint256 underlyingBalance = IERC20(underlyingToken).balanceOf(address(this));
+        if (amount > thisBalance + underlyingBalance) {
+            unchecked {
+                IERC20(underlyingToken).safeTransferFrom(user, address(this), amount - thisBalance - underlyingBalance);
+                underlyingBalance = amount - thisBalance;
+            }
+        }
+        if (underlyingBalance > 0) {
+            IERC20(underlyingToken).approve(token, underlyingBalance);
+            ISubstitute(token).mint(underlyingBalance, address(this));
+        }
+    }
+
+    function _wrapCoupons(Coupon[] memory coupons) internal {
+        // wrap 1155 to 20
+        bytes memory metadata = Wrapped1155MetadataBuilder.buildWrapped1155BatchMetadata(coupons);
+        _couponManager.safeBatchTransferFrom(address(this), address(_wrapped1155Factory), coupons, metadata);
+    }
+
+    function _unwrapCoupons(Coupon[] memory coupons) internal {
+        uint256[] memory tokenIds = new uint256[](coupons.length);
+        uint256[] memory amounts = new uint256[](coupons.length);
+        unchecked {
+            for (uint256 i = 0; i < coupons.length; ++i) {
+                tokenIds[i] = coupons[i].id();
+                amounts[i] = coupons[i].amount;
+            }
+        }
+        bytes memory metadata = Wrapped1155MetadataBuilder.buildWrapped1155BatchMetadata(coupons);
+        _wrapped1155Factory.batchUnwrap(address(_couponManager), tokenIds, amounts, address(this), metadata);
+    }
+
+    function getCouponMarket(CouponKey memory couponKey)
+        external
+        view
+        returns (IBookManager.BookKey memory, IBookManager.BookKey memory)
+    {
+        return (_couponSellMarkets[couponKey.toId()], _couponBuyMarkets[couponKey.toId()]);
+    }
+
+    function setCouponBookKey(
+        CouponKey memory couponKey,
+        IBookManager.BookKey calldata sellBookKey,
+        IBookManager.BookKey calldata buyBookKey
+    ) public virtual onlyOwner {
+        bytes memory metadata = Wrapped1155MetadataBuilder.buildWrapped1155Metadata(couponKey);
+        uint256 couponId = couponKey.toId();
+        address wrappedCoupon = _wrapped1155Factory.getWrapped1155(address(_couponManager), couponId, metadata);
+
+        BookId sellMarketBookId = sellBookKey.toId();
+        BookId buyMarketBookId = buyBookKey.toId();
+        if (
+            _bookManager.getBookKey(sellMarketBookId).unit != sellBookKey.unit
+                || _bookManager.getBookKey(buyMarketBookId).unit != buyBookKey.unit
+        ) {
+            revert InvalidMarket();
+        }
```

Comment:
We should check with wrappedCoupon and couponKey.asset

Reply:
Because `id = key.toId()` we don't need to check `couponKey.asset`

Reply:
Resolved.